### PR TITLE
fix(dvr): fix duplicate recording execution and add manual stop

### LINKED ISSUE #454: [Feature]: Manually stop DVR recording before scheduled end time.
### Is your feature request related to a problem?

My DVR recording for the Cowboys game was set to record from 3:20PM to 7:20 PM, but the game ended (with a pretty bad loss, I might add) by 6:40 PM.  I had to let it continue recording for another 40 minutes even though the game had officially ended.

I also tend to set the DVR to record 30 minutes beyond end time to ensure I capture any long running game.  I'd rather capture extra time if I'm not around to stop it at game end, but it would be nice to stop it manually once complete.

### Describe the solution you'd like

It would be nice to have an option to stop the recording at that point in time and save it as-is.  I would not want the recording deleted as the game itself was captured, just stopped at that point instead of the originally scheduled end time.

### Additional context

Nothing Extra

### LINKED ISSUE #940: [Bug]: A recording starts multiple connections and destroys the recording file
### Describe the bug

I programmed two recordings for two different channels, and according to the DVR view, the recording started right on time at 5:55 p.m. Under “Stats,” I could see that the channel had been activated and five connections had been established. At slightly different times:

5:55:01 p.m.
5:55:19 p.m.
5:55:19 p.m.
5:55:19 p.m.
5:55:19 p.m.

The recordings were supposed to end at 6:30 p.m., but at 6:40 p.m. the recordings were still running according to Stats, even though they were marked as completed under DVR. The file was corrupt at the end. For one entry under DRV, I could press “Watch,” but nothing played, and for the other, it was grayed out.

### How can we recreate this bug?

To recreate my exact case: 2 recordings, same time, different stations

### Dispatcharr Version

v0.18.1

### Docker Version

podman version 5.4.1

### What's in your Docker Compose file?

Default Dispatcharr AIO Configuration

### Client Information

n/a

### PR DESCRIPTION
**Pull request by CodeBormen**
Fixes #940: Duplicate DVR task executions were caused by ClockedSchedule objects being shared across PeriodicTasks, triggering run_recording multiple times for the same recording. Added an idempotency guard in run_recording that exits early if status is already 'recording', 'completed', or 'stopped'.

revoke_task() now deletes the PeriodicTask and orphaned ClockedSchedule on execution rather than relying on Celery revocation alone.

Fixes #454: Added POST /api/channels/recordings/{id}/stop/ action that marks the recording as stopped and terminates only the specific DVR proxy client (identified via User-Agent: Dispatcharr-DVR/recording-{id}) without disrupting simultaneous recordings on the same or other channels. The API returns immediately; DVR client teardown and task revocation happen in a background thread to avoid 504 timeouts on the recordings list endpoint.

Additional changes:
- destroy() now only calls _stop_dvr_clients() for in-progress recordings, preventing accidental stream termination when deleting completed recordings
- WebSocket events differentiated: recording_stopped, recording_cancelled (in-progress cancel), and recording_cancelled (delete) with was_in_progress flag so the frontend shows distinct "Recording deleted" vs "Recording cancelled" notifications

- Frontend: Stop and Cancel split into separate buttons with confirmation modals; stopped recordings move immediately to Previously Recorded via optimistic status bucketing; removed dead 'Stopped' badge state

- 27 new unit tests covering scheduling, idempotency, revocation, and DVR client isolation

### CODE REVIEW

**Comment by CodeBormen** [comment]
## DVR: Fix duplicate recordings (#940) and add manual Stop (#454)

### What was broken

When two recordings were scheduled for the same time, Dispatcharr could start
the same recording multiple times simultaneously. This happened because
ClockedSchedule entries were sometimes shared between PeriodicTasks, firing
`run_recording` more than once per recording. The result was corrupted output
files and recordings that appeared stuck in the "Recording" state long after
they should have completed.

There was also no way to stop a recording early without deleting it entirely — the stop feature allows early recording stops and preserved recording files while also making them eligible for playback and commercial removal

### What's fixed / added

**Duplicate recording fix**
- `run_recording` now exits immediately if the recording is already in
  progress, completed, or stopped — no matter how many times it's called
- `revoke_task()` cleans up both the PeriodicTask and its ClockedSchedule on
  execution, eliminating the orphaned schedule objects that caused duplicates

**Manual Stop (new feature)**
- New **Stop** button (yellow square icon) appears on in-progress recording
  cards alongside the existing Cancel button
- Stop ends the recording early and keeps the partial file available for
  playback; Cancel ends it and deletes everything
- Confirmation modals explain clearly what each action does before you commit
- The Stop API returns immediately — background thread handles stream teardown
  so the UI stays responsive (no more 504 errors on the recordings list)
- When two recordings are running simultaneously, stopping one only terminates
  that recording's proxy connection, leaving the other untouched

**Notification improvements**
- Deleting a completed recording now shows **"Recording deleted"** instead of
  the misleading "Recording cancelled" message
- Stopped recordings move instantly to the Previously Recorded section so
  users don't think the action failed

### Testing

27 automated tests cover: scheduling idempotency, ClockedSchedule cleanup,
task revocation, DVR client isolation (stop one → other keeps running), and
signal guard conditions.

**Comment by CodeBormen** [comment]
Oh also to clarify the first fix, scheduling recordings longer than an hour out was broken because Redis' maximum 3600s timer would hand the job off to another worker. This resulted in multiple workers firing the same job and streaming to the same recording file resulting in corruption. The move to ClockedSchedule prevents this from occurring because the job will be stored in the database and retrieved when needed; it also prevents nonpersistent memory wipes like restarts or upgrades from deleting scheduled jobs.

**Comment by CodeBormen** [comment]
No worries, lots of changes flying around it's all good. Ensuring stability of implemented changes is by far more important.

I'll take a look at the issues you saw and see if I can work through them!

**Comment by CodeBormen** [comment]
Spent the weekend identifying the causes of the issues you mentioned, have been systematically going through and fixing them as well as a few new things and bug testing! Will re-mark for review when ready

**Comment by CodeBormen** [comment]
@SergeantPanda I want to start off by saying that I know this change is fairly large. I fixed quite a few bugs, and when you asked me to go back and look at some of the issues, I kept finding more bugs that lead to more bugs. I thought, "while I'm in here doing this, might as well add some of the things people have been asking for along the way." There are a lot of bug fixes, and a few new features. I really tried to refrain from making any highly opinionated changes, and focused on making things work the way a regular user would expect them to. Spent a lot of time, had some fun and also some of the hardest challenges yet.

If you don't mind reviewing when you have time, I'd appreciate it. I'm happy to take whatever feedback, whether you want additional features added, bugs fixed, or things removed. Don't want it to feel like I am overstepping boundaries.

With that said, here is the list of changes this addresses (so far):

**Recording Reliability**

Fixed duplicate recording execution caused by shared ClockedSchedule objects triggering the same task multiple times (#940)
Added stream reconnection resilience — recordings survive transient network drops with automatic retry (up to 5 attempts), appending to the existing file
Added DB retry with exponential backoff for transient database errors during recording lifecycle
Added crash recovery pipeline — on worker restart, detects recordings stuck in "recording" status, concatenates and remuxes any salvageable content, and revokes stale tasks
Added remux sanity checks to reject suspiciously small output files, preserving the source .ts for manual recovery
Fixed output file collision when multiple tasks target the same filename

**New Features**

Added Stop Recording — cleanly ends an in-progress recording while keeping the partial content playable
Added Extend Recording — extend an in-progress recording by 15m/30m/1h without interrupting the stream
Added inline metadata editing (title/description) in the recording details modal
Added refresh artwork button to re-run poster resolution on demand
Added series rule support for currently-airing episodes (previously only scheduled future ones)

**Cross-Recording Interference**

Fixed send_websocket_update deadlocking the gevent event loop, which caused one recording's events to block another
Stop/cancel now targets only the specific DVR client by recording ID instead of killing all clients on the channel
Artwork & Logos

Fixed recording card logos not displaying (channel summary API shape mismatch)
Added multi-source poster resolution pipeline (EPG, VOD, TMDB, OMDb, TVMaze, iTunes)
Prevented incorrect artwork from fuzzy-matching channel names like "USA A&E SD*" against external APIs
Added negative cache for failed remote logo fetches to stop dead CDNs from blocking Daphne workers

**Series Rules**

Fixed series modal showing "No upcoming episodes" (missing _group_count merge + wrong time filter)
Fixed deleting a series rule leaving orphaned recordings and stale Guide indicators
Fixed recurring rule timezone calculation silently dropping recordings for UTC-negative users after 4pm local (#1042)
Added delete fallback for orphaned recordings when the parent rule no longer exists (#1041)

**Frontend Polish**

Fixed recording modal crash caused by TDZ error in production bundles (editing state used before declaration)
Fixed description textarea losing focus immediately when editing (inline component remount on every render)
Replaced all manual fetchRecordings() calls with debounced WebSocket-driven refresh
Fixed comskip treating "no commercials found" (exit code 1) as a fatal error

**Tests**

Added 168 unit tests across 9 test files covering scheduling, idempotency, stop/cancel, extend, keepalive, DB retry, EPG matching, URL validation, and the recording pipeline

This change would fix/fulfill the following issues/FRs: #940 , #1042 , #1041 , #1038 , #1010 , #1006 , #454 , #619 , #918 , #641 , #473 , #624 , and partially #902

**Comment by SergeantPanda** [comment]
Oh wow this sounds great I'm excited to take a look at how you did it. I will hopefully have time tomorrow or maybe Thursday.

Thank you!

**Review by SergeantPanda — COMMENTED** [review]
Well, I gave it my best effort to read the code, that is a lot of changes!

I did see the changes to the proxy and thought since we added some redis checks here we might want to do some throttling to avoid potential performance impacts. 

I at some point want to go through the proxy code with a fine toothed comb and make it more efficient, that is a problem for another day though.

If you don't agree with the changes, that is OK you can remove them and I wont be angry.

For now, I see you are still working on this so let me know when you'd like me to take another look.

**Comment by CodeBormen** [comment]
Yes, it is quite a lot, apologies. As I discovered issues, I kept having to branch out. But I'm being intentional on this change to not make any changes to databases or migrations since they are already high-risk. 

I'll check out the changes, I'm not perfect so I'm sure they are good additions! 

If you end up having time to build and test this branch, happy to take any feedback you have on changes and build them in. Otherwise, yeah, still making some changes and testing to make sure everything works the way it should. There were a couple changes I made and rolled back just because I don't want to add even more complexity to what is already here.

**Comment by CodeBormen** [comment]
Yes, it is ready for review! I think I am done with any additional changes unless I hear feedback from you about adding/removing anything. I am still playing with it, testing as many edge cases and stress tests as I can to make sure it performs well in all scenarios, and if I find any issues I will fix those.

**Comment by SergeantPanda** [comment]
Could you please look over the changelog for this and verify I documented correctly. I heavily relied on ai for it due to the overwhelming changes to keep track of.

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  - **Stop Recording**: A new Stop button (distinct from Cancel) cleanly ends an in-progress recording early and keeps the partial file available for playback. The API returns immediately; stream teardown and task revocation happen in a background thread to prevent 504 timeouts. When multiple recordings run simultaneously, stopping one only terminates that recording's proxy client by ID, leaving all others unaffected. (Closes #454)
+  - **Extend Recording**: In-progress recordings can be extended by 15, 30, or 60 minutes without interrupting the stream.
+  - **Inline metadata editing**: Title and description can now be edited directly in the recording details modal.
+  - **Refresh artwork button**: Manually re-run poster resolution on demand from the recording card.
+  - **Multi-source poster resolution**: Added pipeline querying EPG, VOD, TMDB, OMDb, TVMaze, and iTunes for richer recording artwork.
+  - **Series rules for currently-airing episodes**: Series rules now capture currently-airing episodes in addition to future scheduled ones. (Closes #473)
+  - **Search and filter controls**: Added search and filter controls to the recordings list.
```
Comment (CodeBormen) [inline CHANGELOG.md:21]: Would add the following below this:

  - **Automatic crash recovery on worker restart**: A `worker_ready` Celery signal now fires `recover_recordings_on_startup` automatically when the worker starts, so recordings stuck in "recording" status are recovered without manual intervention.

File: CHANGELOG.md [hunk 2]
```diff
@@ -31,6 +39,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  - **Recording modal TDZ crash**: Modal crashed on load in production bundles due to a Temporal Dead Zone error — editing state was referenced before its declaration in the minified bundle.
+  - **Description textarea focus loss**: The description textarea lost focus immediately when opened because the inline editing component was remounting on every render.
+  - **WebSocket-driven refresh**: Replaced all manual `fetchRecordings()` polling calls with debounced WebSocket-driven refresh so the recordings list stays up to date without redundant API requests.
+  - **comskip exit code handling**: comskip treated exit code 1 ("no commercials found") as a fatal error, causing post-processing to fail on clean recordings. Exit code 1 is now recognised as a successful no-op.
+  - **Differentiated WebSocket notification events**: `recording_stopped`, `recording_cancelled` (in-progress cancel), and `recording_deleted` with a `was_in_progress` flag now allow the frontend to display distinct "Recording stopped", "Recording cancelled", and "Recording deleted" toasts.
```
Comment (CodeBormen) [inline CHANGELOG.md:62]: Would add the following below this:

  - **Stream generator throttling**: Cached the `ProxyServer` singleton reference per client and throttled Redis resource checks (1 s) and non-owner health checks (2 s), eliminating 3+ Redis round-trips per stream loop iteration.
  - **Duplicate series rule evaluation race**: Creating a series rule fired `evaluate_series_rules.delay()` in the API view while the frontend immediately called the synchronous evaluate endpoint, racing to create duplicate recordings for the same program. Removed the redundant async call from the API; the frontend's explicit evaluate call is now the sole evaluation path.
  - **Recording card S/E badge overlap**: Season/episode badges were overlapping and metadata was hidden on the recording card.
  - **Orphaned recording fallback in series modal**: When a series rule no longer exists, the recurring rule modal now shows a "Delete Recording" button for the orphaned recording instead of failing silently.

File: CHANGELOG.md [hunk 3]
```diff
@@ -31,6 +39,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - Non-atomic ownership acquisition in `try_acquire_ownership()`: replaced the separate `setnx()` + `expire()` calls with a single atomic `SET NX EX`, eliminating the race window where a process crash between the two calls could leave an ownership key with no TTL (permanent ownership lock).
+- DVR bug fixes — Thanks [@CodeBormen](https://github.com/CodeBormen)
+  - **Duplicate recording execution**: Shared `ClockedSchedule` objects were firing `run_recording` multiple times for the same recording, producing corrupted output files. `run_recording` now exits immediately if the recording is already in progress, completed, or stopped (failing closed on DB errors to prevent duplicate tasks from overwriting valid recordings). `revoke_task()` also cleans up both the `PeriodicTask` and its `ClockedSchedule` on execution to eliminate orphaned schedule objects. (Fixes #641)
```
Comment (CodeBormen) [inline CHANGELOG.md]: I would replace lines 43/44 with this one combined change, the changelog as it currently reads suggests that ClockedSchedule was the cause of the duplicate recording issue when it was actually the fix for async causing the duplicates (due to the 3600s issue):

  - **Duplicate recording execution**: `run_recording.apply_async(countdown=...)` exceeded Redis' default `visibility_timeout` (3600 s) for recordings scheduled more than one hour out, causing Redis to redeliver the task to multiple workers simultaneously and producing corrupted output files. Replaced `apply_async` with `ClockedSchedule` + `PeriodicTask` for database-backed one-shot scheduling that survives restarts and upgrades without the redelivery race. `run_recording` also now exits immediately if the recording is already in progress, completed, or stopped. `revoke_task()` cleans up both the `PeriodicTask` and its orphaned `ClockedSchedule` on execution. (Fixes #940, #641)

**Comment by CodeBormen** [comment]
It looks good to me, I did make a few comments on the changelog for a couple additions to what was already there, as well as replacing lines 43/44 which are not entirely accurate to the change

**Comment by SergeantPanda** [comment]
Would you mind just editing it how you see fit? Sorry I just wanted to get something started here to hopefully help with the documentation. 

I normally do it for PR's but this one would be best approved by the submitter.

**Comment by CodeBormen** [comment]
Yep no problem, good to go!

**Comment by SergeantPanda** [comment]
Merged. Very good work on this. We appreciate the time and effort that went into it!
